### PR TITLE
[UI] IssuePage multi repo select

### LINKED ISSUE #6355: [Feature proposal] Multiple repositories selectable @ issues/PR page
## Description

As a user, I would like to be have the option to view issues and pull-requests from multiple repositories on their respective pages. Currently, selecting one repository removes the other repos from the list. This feels unnecessarily limiting, and not very intuitive. My suggestion is to use a similar style, but not remove the repos and also allow multiple selections.

### Possible method
Currently, the page identifies what repo should be used by defining a number of variables including repo (e.g. `?type=all&repo=17&sort=&state=open`). Passing multiple repo entries here could make multiple selection possible, something like 
`?type=all&repos[]=17&repos[]=21&repos[]=22&sort=&state=open`.

I will test out a local fix, and post my findings here.

## Illustration
multiple-repo-select

### PR DESCRIPTION
**Pull request by 6543**
## Continue #6369
close #6355

# Before
all issues | you after klicked to one
--------------|------------------------------
Bildschirmfoto zu 2019-10-30 02-27-20 | Bildschirmfoto zu 2019-10-30 02-27-37

# Avter PR
all issues | you after klicked to one
--------------|------------------------------
Bildschirmfoto zu 2019-10-30 02-32-11 | Bildschirmfoto zu 2019-10-30 02-25-41

### CODE REVIEW

**Comment by 6543** [comment]
`TestXSSUserFullName (integrations/xss_test.go:17)` Fail ?!?

**Comment by 6543** [comment]
@ethantkoenig you added **TestXSSUserFullName** at #2122
do you remember/know why the CI could fail?
[Drone Log](https://drone.gitea.io/go-gitea/gitea/15826/1/13)

**Comment by 6543** [comment]
`template: user/dashboard/issues:90:17: executing "user/dashboard/issues" at <.Repo.FullName>: error calling FullName: runtime error: invalid me
mory address or nil pointer dereference`

**Comment by 6543** [comment]
@lafriks @lunny @kolaente I'll resolfed & fixed the PR from @oscarlofwenhamn 

so I would be happy if you can look it :)

File: models/issue.go [hunk 1]
```diff
@@ -1388,6 +1389,9 @@ func GetUserIssueStats(opts UserIssueStatsOptions) (*IssueStats, error) {
 	if opts.RepoID > 0 {
 		cond = cond.And(builder.Eq{"issue.repo_id": opts.RepoID})
```
Comment (techknowlogick) [inline models/issue.go]: Instead of adding conditional to SQL, could this just append `opts.RepoID` to `opts.RepoIDs` slice?

File: routers/user/home.go [hunk 2]
```diff
@@ -281,46 +290,41 @@ func Issues(ctx *context.Context) {
 
 	showReposMap := make(map[int64]*models.Repository, len(counts))
 	for repoID := range counts {
```
Comment (techknowlogick) [inline routers/user/home.go:297]: should this`repoID` be `repoIDs`?

File: routers/user/home.go [hunk 3]
```diff
@@ -194,7 +196,21 @@ func Issues(ctx *context.Context) {
 		page = 1
 	}
 
-	repoID := ctx.QueryInt64("repo")
+	// Regexp: Match all square brackets, and plus signs in order to handle pagination
+	re := regexp.MustCompile(`[\[\]\+]+`)
```
Comment (techknowlogick) [inline routers/user/home.go]: Should this be compiled each time?

**Review by techknowlogick — DISMISSED** [review]
Haven't gotten to all of this PR, but here are a few notes.

File: routers/user/home.go (hunk 3 above)
Comment (6543) [reply routers/user/home.go]: @techknowlogick is there a special place where RegExp are stored?
if not I'll move it out of this func but leave it in home.go

File: models/issue.go (hunk 1 above)
Comment (6543) [reply models/issue.go]: I checked where UserIssueStatsOptions is used ... we can drop RepoID .... since it is only used in:
 * ./routers/user/home.go: 1time
* ./models/issue.go: 1time
* ./models/issue_test.go: 1time

File: routers/user/home.go (hunk 3 above)
Comment (guillep2k) [reply routers/user/home.go]: > 
> 
> @techknowlogick is there a special place where RegExp are stored?
> if not I'll move it out of this func but leave it in home.go

Check this example:

https://github.com/go-gitea/gitea/blob/8f09aff427add98cbad0254c4849ea6eee34e070/models/issue_label.go#L20

File: models/issue.go (hunk 1 above)
Comment (6543) [reply models/issue.go]: done ...

File: routers/user/home.go (hunk 3 above)
Comment (6543) [reply routers/user/home.go]: done

File: routers/user/home.go (hunk 2 above)
Comment (6543) [reply routers/user/home.go:297]: no - this is a for loop wich do for each repoID ...

**Comment by 6543** [comment]
@guillep2k @techknowlogick this PR is ready to review again :D

File: routers/user/home.go [hunk 4]
```diff
@@ -194,7 +199,19 @@ func Issues(ctx *context.Context) {
...
 	}
 
-	repoID := ctx.QueryInt64("repo")
+	// Replace regexp return value with commas, and split
+	repoIDStrings := strings.Split(issueRepoIDsPattern.ReplaceAllString(ctx.Query("repos"), ","), ",")
+	var repoIDs []int64
+	for _, IDString := range repoIDStrings {
+		// Ensure nonempty string entries
+		if IDString != "" && IDString != "0" {
+			IDint64, err := strconv.ParseInt(IDString, 10, 64)
```
Comment (lunny) [inline routers/user/home.go]: local variable name capital letter should be lower.

File: routers/user/home.go [hunk 5]
```diff
@@ -194,7 +199,19 @@ func Issues(ctx *context.Context) {
 		page = 1
 	}
 
-	repoID := ctx.QueryInt64("repo")
+	// Replace regexp return value with commas, and split
+	repoIDStrings := strings.Split(issueRepoIDsPattern.ReplaceAllString(ctx.Query("repos"), ","), ",")
```
Comment (lunny) [inline routers/user/home.go]: Is this a standard method to request an array?

File: routers/user/home.go [hunk 6]
```diff
@@ -281,46 +291,41 @@ func Issues(ctx *context.Context) {
...
+						"User in repo has Permissions: %-+v",
+						ctxUser,
+						models.UnitTypeIssues,
+						repo,
+						perm)
+				}
+				// ctx.Status(404)
+				// return
+				delete(showReposMap, repoID)
+				delete(counts, repoID)
```
Comment (lunny) [inline routers/user/home.go]: I think you should not delete element of map when traversing it.

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/user/home.go (hunk 5 above)
Comment (6543) [reply routers/user/home.go]: I don't know this part of the code is from @oscarlofwenhamn

I'll have a look at this ...

Comment (oscarlofwenhamn) [reply routers/user/home.go]: Probably not standard, no. I'm afraid it was too long ago since I was working on this, so I don't remember the details of what I was going for that well.
This came about after pagination was introduced to the page in question, which is where I ran into problems. In order to handle the way the pagination operates, I needed to convert the return from `ctx.Query("repos")` into a general format, because I couldn't sort it out in the original format.

Comment (6543) [reply routers/user/home.go]: @oscarlofwenhamn thanks for explaining this ... I rewrote the regex and the loop ...

File: routers/user/home.go (hunk 4 above)
Comment (6543) [reply routers/user/home.go]: done

File: routers/user/home.go (hunk 6 above)
Comment (6543) [reply routers/user/home.go]: done

**Comment by oscarlofwenhamn** [comment]
@6543 Really nice to see this take off, huge thanks for the work and help!

**Comment by 6543** [comment]
@lunny I think i resolved all your cases

**Comment by 6543** [comment]
@techknowlogick can you pleace look at this - thanks

File: routers/user/home.go (hunk 2 above)
Comment (techknowlogick) [reply routers/user/home.go:297]: ah of course. I missed that, thanks :)

File: routers/user/home.go [hunk 7]
```diff
@@ -149,6 +151,9 @@ func Dashboard(ctx *context.Context) {
 	ctx.HTML(200, tplDashboard)
 }
 
+// Regexp for repos query
+var issueReposQueryPattern = regexp.MustCompile(`^\[(\d+,{0,1})+\]$`)
```
Comment (zeripath) [inline routers/user/home.go]:
```suggestion
var issueReposQueryPattern = regexp.MustCompile(`^\[\d+(,\d+)*\]$`)
```

File: routers/user/home.go [hunk 8]
```diff
@@ -194,7 +199,25 @@ func Issues(ctx *context.Context) {
 		page = 1
 	}
 
-	repoID := ctx.QueryInt64("repo")
+	reposQuery := ctx.Query("repos")
+	var repoIDs []int64
+	if issueReposQueryPattern.MatchString(reposQuery) {
+		// remove "[" and "]" from string
+		reposQuery = strings.Replace(strings.Replace(reposQuery, "[", "", 1), "]", "", 1)
```
Comment (zeripath) [inline routers/user/home.go]:
```suggestion
		reposQuery = reposQuery[1 : len(reposQuery)-1]
```

File: routers/user/home.go [hunk 9]
```diff
@@ -194,7 +199,25 @@ func Issues(ctx *context.Context) {
...
+		//for each ID (delimiter ",") add to int to repoIDs
+		for _, rID := range strings.Split(reposQuery, ",") {
+			// Ensure nonempty string entries
+			if rID != "" && rID != "0" {
+				rIDint64, err := strconv.ParseInt(rID, 10, 64)
+				if err == nil {
+					repoIDs = append(repoIDs, rIDint64)
```
Comment (zeripath) [inline routers/user/home.go:213]: I guess it's ok that there are duplicates in this list because our query will just handle these.

File: routers/user/home.go [hunk 10]
```diff
@@ -281,46 +297,37 @@ func Issues(ctx *context.Context) {
...
+			if !perm.CanRead(models.UnitTypeIssues) {
+				if log.IsTrace() {
+					log.Trace("Permission Denied: User %-v cannot read %-v of repo %-v\n"+
+						"User in repo has Permissions: %-+v",
+						ctxUser,
+						models.UnitTypeIssues,
+						repo,
+						perm)
+				}
 			}
```
Comment (zeripath) [inline routers/user/home.go]: How do you account for this read denial?

File: routers/user/home.go (hunk 7 above)
Comment (6543) [reply routers/user/home.go]: there exist two paterns:
`[2,6,3]` and  `[2,6,3,]`

Comment (6543) [reply routers/user/home.go]: this is because of template range methode

Comment (zeripath) [reply routers/user/home.go]: Oh in that case:

```suggestion
var issueReposQueryPattern = regexp.MustCompile(`^\[\d+(,\d+)*,?\]$`)
```

Is still less complex than the original one.

File: routers/user/home.go (hunk 9 above)
Comment (6543) [reply routers/user/home.go:213]: I didn't test it jet but think so ... (testing ...)

Comment (6543) [reply routers/user/home.go:213]: ... later function take care of so dubble IDs is no problem

Comment (oscarlofwenhamn) [reply routers/user/home.go:213]: Yes, this is "by design" (i.e. it could probably be done in another way, but using the later function gives desired result).

**Comment by 6543** [comment]
Nore for me: integration test fail only for specific issue type ... look at this

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
edge case wich cause the problem:

if a user (not admin) creat a issue on a repo he doesnt own,
and this repo is set to private avterwards
it is still count to the issue stats for the user
but thanks to https://github.com/go-gitea/gitea/pull/8741/files#diff-fbb8de13d0c91e636836fe8d3e5baafaR321 you only need one issue of this edge case

**Comment by 6543** [comment]
-> the edge case is not coverd by other fuctions of gitea too!
 * User Activety Page
 * WithOUT this PR it is visible in dash borad
 * more ?

File: routers/user/home.go [hunk 11]
```diff
@@ -281,46 +297,30 @@ func Issues(ctx *context.Context) {
...
-					repo,
-					perm)
+			// Check if user has access to given repository.
+			perm, err := models.GetUserRepoPermission(repo, ctxUser)
+			if err != nil {
+				ctx.ServerError("GetUserRepoPermission", fmt.Errorf("[%d]%v", repoID, err))
+				return
+			}
+			if !perm.CanRead(models.UnitTypeIssues) {
+				log.Error("User created Issues witch he has no longer access to Repo: [%d]", repoID)
```
Comment (zeripath) [inline routers/user/home.go]:
```suggestion
				log.Error("User created Issues in Repository which they no longer have access to: [%d]", repoID)
```

Comment (zeripath) [reply routers/user/home.go]: We still need to following this do:

```go
ctx.Status(404)
return
```

I think?!

Comment (6543) [reply routers/user/home.go]: @zeripath 
I just reproduce the behaviour as bevore -> https://github.com/go-gitea/gitea/issues/9155
If we like to fix this issue we have a litle more "refactor" on this ...

File: routers/user/home.go [hunk 12]
```diff
@@ -281,46 +297,30 @@ func Issues(ctx *context.Context) {
...
-			repo, err := models.GetRepositoryByID(repoID)
-			if models.IsErrRepoNotExist(err) {
-				ctx.NotFound("GetRepositoryByID", err)
-				return
-			} else if err != nil {
-				ctx.ServerError("GetRepositoryByID", fmt.Errorf("[%d]%v", repoID, err))
-				return
+		if repoID > 0 {
+			if _, ok := showReposMap[repoID]; !ok {
+				repo, err := models.GetRepositoryByID(repoID)
```
Comment (guillep2k) [inline routers/user/home.go]:
```suggestion
				showReposMap[repoID], err = models.GetRepositoryByID(repoID)
```

File: templates/user/dashboard/issues.tmpl [hunk 13]
```diff
@@ -5,16 +5,16 @@
 		<div class="ui stackable grid">
 			<div class="four wide column">
 				<div class="ui secondary vertical filter menu">
-					<a class="{{if eq .ViewType "your_repositories"}}ui basic blue button{{end}} item" href="{{.Link}}?type=your_repositories&repo={{.RepoID}}&sort={{$.SortType}}&state={{.State}}">
+					<a class="{{if eq .ViewType "your_repositories"}}ui basic blue button{{end}} item" href="{{.Link}}?type=your_repositories&repos=[]&sort={{$.SortType}}&state={{.State}}">
```
Comment (guillep2k) [inline templates/user/dashboard/issues.tmpl]: Maybe...?
```suggestion
					<a class="{{if eq .ViewType "your_repositories"}}ui basic blue button{{end}} item" href="{{.Link}}?type=your_repositories&repos=[{{range $.RepoIDs}}{{.}}%2C{{end}}]&sort={{$.SortType}}&state={{.State}}">
```

**Comment by 6543** [comment]
@guillep2k done + made some tweks

**Review by guillep2k — APPROVED** [review]

**Comment by 6543** [comment]
ready for merge :rocket:

**Comment by 6543** [comment]
@techknowlogick are you ok with this PR?

**Comment by 6543** [comment]
l

**Comment by 6543** [comment]
ready for merge :rocket:

**Comment by 6543** [comment]
:sweat_smile: we got to the 100 commits :see_no_evil:
